Add "Import recovery keys" button to Payments->Advanced->Recover

willy-b opened the pull request:
Add "Import recovery keys" button that opens a file chooser dialog

fixes #4806
fixes #6263

- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist. (issue #4806)
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed). 

## Test Plan:
1. Open Preferences->Payments

-- backup your wallet to a file  --
2. Click "Advanced Settings..."
3. Click "Backup your wallet"
4. Click "Save recovery file..."
5. Choose a location / filename for your backup from the File Chooser dialog

-- recover your wallet from file --
6. Return to Preferences -> Payments (Close all modals)
7. Open "Advanced Settings..."
8. Click "Recover your wallet"
9. Click "Import recovery keys"
10. Use File Chooser dialog to select the backup file created previously (step 5)
11. Observe "Success" screen

## Automated test plan
```bash
# window 1
npm run watch-test
#window 2
npm run uitest -- --grep="Payments Panel -> Advanced Panel"
```

willy-b commented:
cc/ @luixxiul

bbondy commented:
Would you mind rebasing this on master? Thanks!

willy-b commented:
@bbondy: no problem, rebased against latest

bsclifton commented:
@willy-b doh- needs another rebase.  If you can get that done, I'll make sure to review in a timely manner

willy-b commented:
@bsclifton , rebasing and adjusting for the upstream changes in the Recovery Key File format now

willy-b commented:
OK @bsclifton , rebase is done (old branch copied to https://github.com/willy-b/browser-laptop/tree/add-import-recovery-keys-button-4806-rebase-old ). 

Thanks!

bsclifton commented:
Some notes after reviewing:

- When choosing a recovery file and you hit cancel, it shows the error screen. **Could we change this to not show an error and expect the user to click the "import recovery keys" button again?**
- When importing is a success, it shows a nicely formatted message showing the balance, etc.  However, it stays on the Advanced Settings screen. When user clicks the "OK" button (and it's a success), **can we close all the modals so that they're returned to the regular Brave Payments screen?**
- The format of the recovery keys file which is read doesn't appear to be checked which can lead to folks falsely thinking they've restored properly. **Can we validate the contents of the file or give a better response?** For example, I backed up my wallet and then edited the text file; these scenarios all give a success message saying 0 BTC was imported:
  - removing parts of one or both recovery keys
  - removing the 2nd recovery key
  - removing both of the recovery
  - deleting all the contents of the file

willy-b commented:
Don't necessarily re-review yet, one more commit coming! This commit just contains tests for saving and recovering the key file which I wanted to keep distinct. Thanks.

willy-b commented:
OK @bsclifton, this should be ready for re-review. Everything in your previous comment should be fixed now. After review I'll do a rebase and force-push if it's OK with you.

posix4e commented:
Any reason why we can't clear this out? I thin kthis already works?

bsclifton commented:
@posix4e I haven't had time to review :frowning_face: If you have some time and can try this out (needs a rebase), I'd say go ahead and try it :smile:

posix4e commented:
@bsclifton roger that

posix4e commented:
Actually @bsclifton  not sure what this commit does that we don't already @willy-b

willy-b commented:
Hey @posix4e, thanks for taking a look.

Re your question, there's a few commits here: 
- If you're coming from #6263 , there's a single commit that can be cherry-picked to fix that (close the Advanced and Recover popups when importing keys successfully) -> https://github.com/brave/browser-laptop/pull/5809/commits/e01ad045f0c6074377c3cf471d1a4c2430ea3642

- The main feature of this  PR is adding a button to allow import of the recovery file, so the user doesn't have to parse it and enter the keys manually (originally requested in #4806).

- This PR also adds the first tests for the existing recovery keys backup feature, afaik

posix4e commented:
@willy-b wanna rebase then?

willy-b commented:
Sure, rebasing now (just pushed a rebase but it was to an out of date master).

The test command will be:
- just the new tests for the Payments -> Advanced modals:
`npm run test -- --grep="Payments Panel -> Advanced Panel"`

- all Payments Panel tests to look for regressions:
`npm run test -- --grep="Payments Panel"`

willy-b commented:
Rebased against latest master, thanks.

bsclifton commented:
@willy-b can you update the commit that fixes https://github.com/brave/browser-laptop/issues/6263 to have the "Fixes https://github.com/brave/browser-laptop/issues/6263" keyword in the commit? :smile: 

Sorry for the extremely long turn-around here. I know we're all hustling on the open items for 0.13.0, which will be our first release with Chromium 54 (which has caused us to have major architectural changes). This is definitely an important fix and I'm excited about pulling it (and the other PRs for 0.13.1) in.  Just wish I was making better progress with some of the items I'm stuck on :frowning_face:

willy-b commented:
@bsclifton, added the ref to #6263 in commit https://github.com/willy-b/browser-laptop/commit/f1d9e52265bd6cffd83af81d6ee69ef281ead9b2 and rebased against latest master

willy-b commented:
note: just saw that #6422 is about to come in; on review I see that this PR will need an update after https://github.com/brave/browser-laptop/pull/6422/commits/a6cc6a58ab773c1e165b83f3b021a0a4fea65f38 is merged.

so after #6422 is merged let me update this one last time before this PR is merged or it will break, thanks.

willy-b commented:
heads up, 2 more pending PRs which may cause silent conflict when merged, bringing total to 3:
- https://github.com/brave/browser-laptop/pull/6455 (new)
- https://github.com/brave/browser-laptop/pull/6456 (new)
- https://github.com/brave/browser-laptop/pull/6422

(if these are merged before this is, hold this until I or someone else tests it again. otherwise I think this is clear)

luixxiul commented:
Is this ready-for-merge, or do we need another review?

willy-b commented:
This is ready-to-merge, unless #6455, #6456, or #6422 is merged first.

bsclifton commented:
@luixxiul I haven't reviewed it; with all of the open issues, it doesn't make sense to continue spending time reviewing PRs (instead of fixing the open issues with 0.13.0).  I'll be 100% sure to revisit 0.13.1 items after solving open items with 0.13.0 :smile:

luixxiul commented:
There is another conflict to be fixed.

willy-b commented:
thanks @luixxiul, looks like some conflicting changes finally came in since last comment. I'll resolve that today!

willy-b commented:
looks like you just ebased it and resolved conflicts yourself @bsclifton. thanks!
beat me to it -- was still testing locally, have some test issues with latest master.

I'll pull and test what you've got here then.

willy-b commented:
thanks @bsclifton!
I'm trying to test this but as was the case yesterday I continue to have build / run issues with latest master (e.g. webpack server turns off after I run the app, so end up with a blank window and localhost:8080 refused errors in console). I'll update you once that is resolved and I can test your changes.

willy-b commented:
FYI, #6455 just came in so may be silent conflict with that.

bsclifton commented:
@willy-b yup- understood, as you mentioned herehttps://github.com/brave/browser-laptop/issues/6447#issuecomment-269571966

willy-b commented:
hey @bsclifton, I couldn't get that branch to build for whatever reason -- probably a local issue.
Unfortunately I can't commit to review this until this weekend. Feel free to re-assign or do what you have to do.

bsclifton commented:
@willy-b no review needed :smile: I spent a good amount of time with it; we've changed a few things about our PRs which may be confusing (especially since I am not sure we documented it!)

- Assignee for the PR will be the folks who are doing the change OR own the change
- Reviewers would be the folks who do the actual review in GitHub and approve/deny

I'll make sure our docs get updated.  Thanks again; great work on this :smile:

luixxiul commented:
Another rebase is needed and [7 failing tests](https://travis-ci.org/brave/browser-laptop/builds/190718419#L4559) need passing to be merged with 0.13.1-branch.

bsclifton commented:
[CI build had 3 errors](https://travis-ci.org/brave/browser-laptop/builds/191213665); I ran each locally (one being a group of 6) and they each passed great.  I removed the commit I had done to increase the timeout (it didn't improve things).

Changes are good to go; merging to `0.13.1-branch` :smile:

luixxiul commented:
Nice work @willy-b :+1:
